### PR TITLE
TYPING: more type hints for io.common

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -126,9 +127,17 @@ def _validate_header_arg(header) -> None:
         )
 
 
-def _stringify_path(
-    filepath_or_buffer: FilePathOrBuffer[AnyStr]
-) -> FilePathOrBuffer[AnyStr]:
+@overload
```
Comment (jreback) [inline pandas/io/common.py]: https://stackoverflow.com/questions/37553551/3-5-type-hinting-and-method-overloading

is there a reason we *need* overload here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/common.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/io/common.py]: we could be less strict and use casts or ignores instead where necessary

Comment (jreback) [reply pandas/io/common.py]: this seems to be hiding a bug then, these signatures *should* all be the same no? why do we *need* different signatures?

Comment (simonjayhawkins) [reply pandas/io/common.py]: the signatures are the same. the types are different.

buf passed through unchanged (as typevar)

Path converted to string

so overload is so that if a Path is given a string is returned.

Comment (jreback) [reply pandas/io/common.py]: I see so this is essentially doing a Union of 2 TypeVars then. ok, can you give a couple of comments before the use of overload then to explain what is happening. my hesitation is that the next person to look at this will not understand w/o some hints.

Comment (WillAyd) [reply pandas/io/common.py]: This is a great application of overload - thanks @simonjayhawkins

File: pandas/io/common.py [hunk 2]
```diff
@@ -126,9 +127,17 @@ def _validate_header_arg(header) -> None:
...
+def _stringify_path(filepath_or_buffer: Union[str, Path]) -> str:
+    ...
```
Comment (jbrockmendel) [inline pandas/io/common.py]: Does ellipses vs pass matter here?  AFAIK could be “overload” magic

Comment (WillAyd) [reply pandas/io/common.py]: I think this is just a typing convention:

https://mypy.readthedocs.io/en/latest/stubs.html#stub-file-syntax

^ noted for stub files but I think OK here as well

File: pandas/io/common.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/io/common.py]: To see if I understand correctly, overloading allows us to properly type e.g.:

```
def foo(x, y):
   if isinstance(x, Foo) and isinstnace(y, Bar):
        [do stuff]
   elif isinstance(x, Bar) and isinstnace(y, Foo):
        [do stuff]
   else:
        raise TypeError(x, y)
```

where if we just typed it as `Union[Foo, Bar]` for both `x` and `y` mypy would infer that Bar, Bar was OK when it isn't.  And overload allows us to specify the exact combinations that are OK.

Is that close?

File: pandas/io/common.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/io/common.py]: see https://docs.python.org/3/library/typing.html#typing.overload for example.

File: pandas/io/common.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/io/common.py]: yep.

Comment (simonjayhawkins) [reply pandas/io/common.py]:
```python
from typing import overload


class Foo:
    pass


class Bar:
    pass


@overload
def foo(x: Foo, y: Bar) -> str:
    ...


@overload
def foo(x: Bar, y: Foo) -> int:
    ...


def foo(x, y):
    if isinstance(x, Foo) and isinstance(y, Bar):
        return "foo"
    elif isinstance(x, Bar) and isinstance(y, Foo):
        return 42
    else:
        raise TypeError(x, y)


foo(Foo(), Bar())
foo(Bar(), Foo())
foo(Foo(), Foo())
foo(Bar(), Bar())
```
```
$ mypy test.py
test.py:33: error: No overload variant of "foo" matches argument types "Foo", "Foo"
test.py:33: note: Possible overload variants:
test.py:33: note:     def foo(x: Foo, y: Bar) -> str
test.py:33: note:     def foo(x: Bar, y: Foo) -> int
test.py:34: error: No overload variant of "foo" matches argument types "Bar", "Bar"
test.py:34: note: Possible overload variants:
test.py:34: note:     def foo(x: Foo, y: Bar) -> str
test.py:34: note:     def foo(x: Bar, y: Foo) -> int
```

Comment (simonjayhawkins) [reply pandas/io/common.py]: > I see so this is essentially doing a Union of 2 TypeVars then. ok, can you give a couple of comments before the use of overload then to explain what is happening. my hesitation is that the next person to look at this will not understand w/o some hints.

@jreback I imagine we could be using overload quite a bit. I'll add comments if you think necessary, but it may end up adding quite a bit of duplication with the docstrings over time.

Comment (jreback) [reply pandas/io/common.py]: maybe add a small section in the contributing docs then of why / how we are using it?

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -458,7 +458,7 @@ def _read(filepath_or_buffer: FilePathOrBuffer, kwds):
     finally:
         parser.close()
 
-    if should_close:
+    if should_close and not isinstance(fp_or_buf, str):
```
Comment (jreback) [inline pandas/io/parsers.py]: why is this here and not correctly coming from get_filepath_or_buffer?

File: pandas/io/common.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/io/common.py]: from https://mypy.readthedocs.io/en/latest/more_types.html#runtime-behavior...

"While we can leave the variant body empty using the pass keyword, the more common convention is to instead use the ellipsis (...) literal."

@jbrockmendel mark as resolved?

File: pandas/io/common.py [hunk 4]
```diff
@@ -126,9 +127,17 @@ def _validate_header_arg(header) -> None:
...
+
+
+@overload
+def _stringify_path(filepath_or_buffer: IO[AnyStr]) -> IO[AnyStr]:
```
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/io/common.py]: i think we will want to add explicit type parameters to all generics as more type hints are added so that we may add `--disallow-any-generics` to allow stricter checking in the future.

File: pandas/io/common.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/io/common.py]: Contributing Guide for Type Hints #27050 is not yet merged. added comments for now using comments from https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading as boilerplate.

File: pandas/io/parsers.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/io/parsers.py]: `get_filepath_or_buffer` can return a str.

as stated before, my preference would be to always add `#type:ignore` rather than change code unnecessarily. The code is currently type safe due to the try/except block, but mypy doesn't see that.

without the isinstance check, `pandas\io\parsers.py:463: error: Item "str" of "Union[str, IO[Any]]" has no attribute "close"`

Comment (simonjayhawkins) [reply pandas/io/parsers.py]: > The code is currently type safe due to the try/except block, but mypy doesn't see that.

sorry i'm wrong here.

it would be type safe if we used `except (AttributeError, ValueError)` but mypy would still report the same error and would need either an ignore or the isinstance check to pass.

at the moment `should_close` is never `True` when `fp_or_buf` is a string. but then again the purpose of type checking is to prevent bugs that may be added in the future.

maybe `get_filepath_or_buffer` should be converted to a context manager and handle the closing.

File: pandas/io/common.py [hunk 5]
```diff
@@ -128,9 +129,33 @@ def _validate_header_arg(header) -> None:
...
+    ...
+
+
+# The actual *implementation* of '_stringify_path'.
```
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/io/common.py]: The comment is copy and pasted from https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading

When adding the overloads I removed the type annotations from the actual implementation to avoid duplication.

I guess this was the wrong thing to do. will add them back.
